### PR TITLE
Add new cross-platform APIs to Jsrt

### PR DESCRIPTION
**Pull request by digitalinfinity**
This change introduces the cross-platform API shape for JSRT.
It starts with the following assumptions:
-  Changing the definition of existing APIs is ok, as long as the effective shape doesnt change
   (so it's fine to use new macros etc as long as signature/calling convention/linkage etc doesn't change)
-  Existing APIs which take wchar_t* as a parameter will be exposed only on Windows
-  New variants of these APIs that take in UTF-8 strings will be exposed on *all* platforms

Here is a summary of the changes:

New macros introduced:
CHAKRA_CALLBACK - Calling convention attribute of callback functions consumed by jsrt APIs. Used in place of Win32 CALLBACK macro.
CHAKRA_API - Calling convention, linkage and return type of Jsrt APIs (always returns JsErrorCode). Used in place of Win32 STDAPI_ macro.

New types introduced
ChakraCookie - Type of cookies consumed by Jsrt Apis. Used in place of Win32 DWORD_PTR datatype.
JsSerializedScriptLoadUtf8SourceCallback - Callback called to load source code of serialized script encoded as utf8

Following APIs become WIN32 only:
JsParseScript
JsParseScriptWithFlags
JsRunScript
JsExperimentalApiRunModule
JsSerializeScript
JsParseSerializedScriptWithCallback
JsRunSerializedScriptWithCallback
JsParseSerializedScript
JsRunSerializedScript
JsGetPropertyIdFromName
JsGetPropertyNameFromId
JsPointerToString
JsStringToPointer

New APIs introduced (available on all platforms):
JsParseScriptUtf8
JsParseScriptWithAttributesUtf8
JsRunScriptUtf8
JsExperimentalApiRunModuleUtf8
JsRunScriptUtf8
JsExperimentalApiRunModuleUtf8
JsSerializeScriptUtf8
JsParseSerializedScriptUtf8
JsRunSerializedScriptUtf8
JsGetPropertyIdFromNameUtf8
JsGetPropertyNameFromIdUtf8
JsPointerToStringUtf8
JsStringToPointerUtf8

Couple notes:
-      All utf8 strings are passed in as char*
-      All string parameters in the new APIs are utf8 (e.g source, url etc).

### CODE REVIEW

**Comment by digitalinfinity** [comment]
@dilijev fixed the whitespace issue. @agarwal-sandeep @ianwjhalliday any comments? I think @curtisman @Yongqu @liminzhu have already reviewed so if you guys are fine with this, I'll merge to master

File: lib/Jsrt/ChakraCommon.h [hunk 1]
```diff
@@ -23,8 +23,49 @@
...
+#define CHAKRA_CALLBACK __attribute__((cdecl))
+#else
+#define CHAKRA_CALLBACK
+#endif
+
+#define CHAKRA_API extern "C" JsErrorCode
+
+typedef uintptr_t ChakraCookie;
+typedef unsigned char* ChakraBytePtr;
+#endif
```
Comment (agarwal-sandeep) [inline lib/Jsrt/ChakraCommon.h]: Can we add comments at all #else and #endif to mark which #if it refers to? makes it easier to read the definition blocks

File: lib/Jsrt/ChakraCommon.h [hunk 2]
```diff
@@ -1114,10 +1168,247 @@
     /// <returns>
     ///     The code <c>JsNoError</c> if the operation succeeded, a failure code otherwise.
     /// </returns>
-    STDAPI_(JsErrorCode)
+    CHAKRA_API
         JsGetPropertyNameFromId(
             _In_ JsPropertyIdRef propertyId,
             _Outptr_result_z_ const wchar_t **name);
+#endif
```
Comment (agarwal-sandeep) [inline lib/Jsrt/ChakraCommon.h]: comment which #if it ends? _WIN32?

File: lib/Jsrt/ChakraCommon.h [hunk 3]
```diff
@@ -1505,7 +1797,7 @@
     /// <returns>
     ///     The code <c>JsNoError</c> if the operation succeeded, a failure code otherwise.
     /// </returns>
-    STDAPI_(JsErrorCode)
+    CHAKRA_API
         JsPointerToString(
```
Comment (agarwal-sandeep) [inline lib/Jsrt/ChakraCommon.h]: Why creating a new #ifdef _WIN32 block, should we move these 2 functions in the above #ifdef _WIN32 block?

File: lib/Jsrt/ChakraCommon.h [hunk 4]
```diff
@@ -1588,7 +1926,7 @@
     /// <param name="data">
     ///     The external data that was passed in when creating the object being finalized.
     /// </param>
-    typedef void (CALLBACK *JsFinalizeCallback)(_In_opt_ void *data);
+    typedef void (CHAKRA_CALLBACK *JsFinalizeCallback)(_In_opt_ void *data);
```
Comment (agarwal-sandeep) [inline lib/Jsrt/ChakraCommon.h]: typedef enum _JsPropertyIdType
typedef enum _JsValueType
CHAKRA_CALLBACK *JsFinalizeCallback
CHAKRA_CALLBACK * JsNativeFunction
CHAKRA_CALLBACK *JsPromiseContinuationCallback

should be moved to top before any function deceleration to be consistent

**Comment by digitalinfinity** [comment]
Thanks @agarwal-sandeep. Updated the PR.

**Comment by agarwal-sandeep** [comment]
LGTM

**Comment by ianwjhalliday** [comment]
:+1:

**Comment by dilijev** [comment]
@agarwal-sandeep Do you know why the majority of ChakraCommon.h is indented without a block wrapping it? (Starting at line 69)

**Comment by dilijev** [comment]
LGTM

**Comment by agarwal-sandeep** [comment]
@dilijev it has been like this when we split jsrt.h for USE_EDGEMODE_JSRT

**Comment by digitalinfinity** [comment]
Thanks all- going to merge this guy to master

**Comment by kivancmuslu** [comment]
> 
This is a test comment for CodeFlow, please ignore.

---
Refers to: lib/Jsrt/ChakraCommon.h:69 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> #define CHAKRA_API extern "C" JsErrorCode 
This is a CodeFlow test comment, please ignore.

---
Refers to: lib/Jsrt/ChakraCommon.h:69 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> #define CHAKRA_API extern "C" JsErrorCode 

This is a CodeFlow test comment. Please ignore.

---
Refers to: lib/Jsrt/ChakraCommon.h:69 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> #define CHAKRA_API extern "C" JsErrorCode 

This is comment for testing reply, please ignore.

---
In reply to: [207689366](https://github.com/Microsoft/ChakraCore/pull/766#issuecomment-207689366) [](ancestors = 207689366)

---
Refers to: lib/Jsrt/ChakraCommon.h:69 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
>     ///     An error code returned from a Chakra hosting API.

This is a test comment on an unchanged line. Please ignore.

---
Refers to: lib/Jsrt/ChakraCommon.h:79 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
>     typedef _Return_type_success_(return == 0) enum _JsErrorCode : unsigned int

This is a test comment on a deleted line. Please ignore.

---
Refers to: lib/Jsrt/ChakraCommon.h:33 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = True)

**Comment by kivancmuslu** [comment]
> #define CHAKRA_API extern "C" JsErrorCode 

This is a test, please ignore.

---
Refers to: lib/Jsrt/ChakraCommon.h:69 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> test/benchmarks/*.dpl

This is a test for commenting outside the patch.

---
Refers to: .gitignore:31 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> test/benchmarks/*.dpl

Testing a reply to the comment outside the patch. Please ignore.

---
In reply to: [207693118](https://github.com/Microsoft/ChakraCore/pull/766#issuecomment-207693118) [](ancestors = 207693118)

---
Refers to: .gitignore:31 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> #define CHAKRA_API extern "C" JsErrorCode 

Testing reply on no GitHub path. Please ignore.

---
In reply to: [207692300](https://github.com/Microsoft/ChakraCore/pull/766#issuecomment-207692300) [](ancestors = 207692300)

---
Refers to: lib/Jsrt/ChakraCommon.h:69 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
>     typedef _Return_type_success_(return == 0) enum _JsErrorCode : unsigned int

Testing comment on a deleted line. Please ignore.

---
Refers to: lib/Jsrt/ChakraCommon.h:33 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = True)

**Comment by kivancmuslu** [comment]
>     typedef _Return_type_success_(return == 0) enum _JsErrorCode : unsigned int

Testing reply on a comment created without restart. Please ignore.

---
In reply to: [207693897](https://github.com/Microsoft/ChakraCore/pull/766#issuecomment-207693897) [](ancestors = 207693897)

---
Refers to: lib/Jsrt/ChakraCommon.h:33 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = True)

**Comment by kivancmuslu** [comment]
> 

This is a test comment, please ignore.

---
Refers to: .gitignore:32 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> 

This is a test reply, please ignore.

---
In reply to: [207695787](https://github.com/Microsoft/ChakraCore/pull/766#issuecomment-207695787) [](ancestors = 207695787)

---
Refers to: .gitignore:32 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> test/benchmarks/*.dpl

Test comment, please ignore.

---
Refers to: .gitignore:31 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> test/benchmarks/*.dpl

Test reply, please ignore.

---
In reply to: [207695863](https://github.com/Microsoft/ChakraCore/pull/766#issuecomment-207695863) [](ancestors = 207695863)

---
Refers to: .gitignore:31 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> #if defined(_WIN32) && defined(_MSC_VER)

Test comment, please ignore.

---
Refers to: lib/Jsrt/ChakraCommon.h:27 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> #if defined(_WIN32) && defined(_MSC_VER)

Test reply, please ignore.

---
In reply to: [207695890](https://github.com/Microsoft/ChakraCore/pull/766#issuecomment-207695890) [](ancestors = 207695890)

---
Refers to: lib/Jsrt/ChakraCommon.h:27 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> #ifndef _CHAKRACOMMON_H_

Test comment. Please ignore.

---
Refers to: lib/Jsrt/ChakraCommon.h:23 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> #ifndef _CHAKRACOMMON_H_

Test reply, please ignore.

---
In reply to: [207699214](https://github.com/Microsoft/ChakraCore/pull/766#issuecomment-207699214) [](ancestors = 207699214)

---
Refers to: lib/Jsrt/ChakraCommon.h:23 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> test/benchmarks/*.dpl

Test comment, please ignore.

---
Refers to: .gitignore:31 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Comment by kivancmuslu** [comment]
> test/benchmarks/*.dpl

Test reply after restart. Please ignore.

---
In reply to: [207699231](https://github.com/Microsoft/ChakraCore/pull/766#issuecomment-207699231) [](ancestors = 207699231)

---
Refers to: .gitignore:31 in 5c74acc. [](commit_id = 5c74acc466f00222b84392130220592dd65a82c3, deletion_comment = False)

**Pull request merged**